engine/verification/fetcher/execution_fork_test.go

```diff
@@ -0,0 +1,171 @@
+package fetcher_test
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+
+	verificationtest "github.com/onflow/flow-go/engine/verification/utils/unittest"
+	"github.com/onflow/flow-go/model/chunks"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/verification"
+	storage "github.com/onflow/flow-go/storage/mock"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+// TestExecutionForkWithDuplicateAssignedChunks evaluates behavior of fetcher engine respect to
+// receiving duplicate identical assigned chunks on execution forks, i.e., an execution fork with two distinct results, where
+// first chunk of both execution results are the same (i.e., duplicate), and both duplicate chunks are assigned to this
+// verification node.
+//
+// The test asserts that a chunk data pack is requested for those duplicate chunks, and
+// once the chunk data pack arrives, a verifiable chunk is shaped for each of those duplicate chunks, and
+// verifiable chunks are pushed to verifier engine.
+func TestExecutionForkWithDuplicateAssignedChunks(t *testing.T) {
+	s := setupTest()
+	e := newFetcherEngine(s)
+
+	// resultsA and resultB belong to an execution fork and their first chunk (statusA and statusB)
+	// is duplicate and assigned to this verification node.
+	block, resultA, statusA, resultB, statusB, collMap := executionResultForkFixture(t)
+	assignedChunkStatuses := verification.ChunkStatusList{statusA, statusB}
+
+	// executorsA and executorsB are execution node identities that executed resultA and resultB, respectively.
+	_, _, executorsA, executorsB := mockReceiptsBlockIDForConflictingResults(t, block.ID(), s.receipts, resultA, resultB)
+	s.metrics.On("OnAssignedChunkReceivedAtFetcher").Return().Times(2)
+	mockStateAtBlockIDForIdentities(s.state, block.ID(), executorsA.Union(executorsB))
+
+	// the chunks belong to an unsealed block, so their chunk data pack is requested.
+	mockBlockSealingStatus(s.state, s.headers, block, false)
+
+	// mocks resources on fetcher engine side.
+	mockResultsByIDs(s.results, []*flow.ExecutionResult{resultA, resultB})
+	mockBlocksStorage(s.blocks, s.headers, block)
+	mockPendingChunksAdd(t, s.pendingChunks, assignedChunkStatuses, true)
+	mockPendingChunksRem(t, s.pendingChunks, assignedChunkStatuses, true)
+	mockPendingChunksGet(s.pendingChunks, assignedChunkStatuses)
+
+	// fetcher engine must create a chunk data request for each of chunk statusA and statusB
+	requestA := chunkRequestFixture(resultA.ID(), statusA, executorsA, executorsB)
+	requestB := chunkRequestFixture(resultB.ID(), statusB, executorsB, executorsA)
```

Comment:
So in the execution fork case, we send 2 CHDP requests, but only process one of the responses? (see previous comment in `requester.go`)

Reply:
This test only accounts for execution forks with duplicate chunks, e.g., `result1` and `result2` are execution results of the same block, and `result1.ID()` is different than `result2.ID()`. However, both results have the same chunk at their index 0, i.e., `result1.Chunks[0] = result2.Chunks[0]`.

It also assumes that both chunks are assigned to this verification node and are processed concurrently.

For chunk index 0, since both results have the same chunk, and hence the same chunk ID, and since chunk data packs are associated with chunk IDs, we send two chunk requests to the requester engine. However, requester engine that receives these two chunk requests concurrently, realizes that they both belong to the same chunk ID, hence only requests one chunk data pack for that chunk ID. 

When requester engine receives the chunk data pack back, it sends that chunk data pack twice to the fetcher engine: once for `resultA.Chunks[0]` and once for `resultB.Chunks[0]`. The reason is each of these chunks take their own verification path and literally two result approvals will be generated: one for chunk index 0 of `resultA` and the other one for chunk index 0 or `resultB`. 

Note that if these two duplicate chunks are coming sequentially at the non-overlapping intervals, the requester will treat them distinct, i.e., a chunk data pack is requested for each of them. Once a chunk data pack response is delivered to fetcher engine, the requester engine cleans the memory about it, so any further duplicate requests will also be processed. However, this test covers the edge case of concurrently processing duplicate chunks of an execution forks.